### ISSUE TITLE
Issue with switching scenes

### ISSUE THREAD

**Issue by quangminhdinh** [homuler/MediaPipeUnityPlugin#251]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 20.04, WSL2): Windows
- Target (e.g. desktop cpu, android arm64): android arm64
- Bazel version: 4.0.0
- Python version: 3.9
- GCC/G++ version: 11.1.0
- Unity version: 2020.3.8f1
- Android NDK version (if building for Android): 4.9
- Xcode version (if building for iOS):

**Describe the problem**
The camera was working fine after being launched on android. However, after I switched to another scene and returned, the camera stopped working and show the error.

**Full logs**
setView = com.android.internal.policy.DecorView@5e69823 TM=true
Relayout returned: old=(0,0,1080,2280) new=(0,108,1080,2280) req=(1080,2280)0 dur=22 res=0x7 s={true 508245616448} ch=true fn=-1
[DP] dp(1) 1 android.view.ViewRootImpl.reportNextDraw:10957 android.view.ViewRootImpl.performTraversals:3845 android.view.ViewRootImpl.doTraversal:2618 
[DP] dp(2) 1 android.view.SurfaceView.updateSurface:1311 android.view.SurfaceView$1.onPreDraw:225 android.view.ViewTreeObserver.dispatchOnPreDraw:1124 
[DP] pdf(1) 1 android.view.SurfaceView.notifyDrawFinished:577 android.view.SurfaceView.performDrawFinished:564 android.view.SurfaceView.lambda$TWz4D2u33ZlAmRtgKzbqqDue3iM:0 
[DP] pdf(0) 1 android.view.ViewRootImpl.performDraw:4809 android.view.ViewRootImpl.performTraversals:3866 android.view.ViewRootImpl.doTraversal:2618 
[DP] rdf()
MSG_WINDOW_FOCUS_CHANGED 1 1
MemoryManager: Using 'Dynamic Heap' Allocator.
Relayout returned: old=(0,108,1080,2280) new=(0,108,1080,2280) req=(1080,2172)0 dur=13 res=0x1 s={true 508245616448} ch=false fn=2
updateBoundsLayer: shouldReparent = false t = android.view.SurfaceControl$Transaction@1ee2f80 sc = Surface(name=Bounds for - com.yitec.tryon/com.unity3d.player.UnityPlayerActivity@0)/@0x1e369b9 frame = 2
SystemInfo CPU = ARM64, Cores = 8, Memory = 7423mb

SystemInfo ARM big.LITTLE configuration: 2 big (mask: 0xc0), 6 little (mask: 0x3f)

ApplicationInfo com.yitec.tryon version 0.1 build d587582f-aa1f-4f1a-b48f-20685f9d0a8a

Built from '2020.3/staging' branch, Version '2020.3.8f1 (507919d4fff5)', Build type 'Release', Scripting Backend 'il2cpp', CPU 'arm64-v8a', Stripping 'Disabled'

Relayout returned: old=(0,108,1080,2280) new=(0,108,1080,2280) req=(1080,2172)0 dur=29 res=0x1 s={true 508245616448} ch=false fn=3
[DP] dp(1) 1 android.view.SurfaceView.updateSurface:1311 android.view.SurfaceView.setFrame:646 android.view.View.layout:24472 
[DP] pdf(0) 1 android.view.SurfaceView.notifyDrawFinished:577 android.view.SurfaceView.performDrawFinished:564 android.view.SurfaceView.lambda$TWz4D2u33ZlAmRtgKzbqqDue3iM:0 
[DP] rdf()
MSG_RESIZED_REPORT: frame=(0,108,1080,2280) ci=(0,0,0,0) vi=(0,0,0,0) or=1
[DP] dp(1) 1 android.view.ViewRootImpl.reportNextDraw:10957 android.view.ViewRootImpl.access$1200:256 android.view.ViewRootImpl$ViewRootHandler.handleMessage:6101 
Company Name: Yitec
Product Name: TryOn
Relayout returned: old=(0,108,1080,2280) new=(0,108,1080,2280) req=(1080,2172)0 dur=24 res=0x1 s={true 508245616448} ch=false fn=5
[DP] pdf(0) 1 android.view.ViewRootImpl.performDraw:4809 android.view.ViewRootImpl.performTraversals:3866 android.view.ViewRootImpl.doTraversal:2618 
[DP] rdf()
Relayout returned: old=(0,108,1080,2280) new=(0,0,1080,2280) req=(1080,2172)0 dur=20 res=0x21 s={true 508245616448} ch=false fn=8
updateBoundsLayer: shouldReparent = false t = android.view.SurfaceControl$Transaction@1ee2f80 sc = Surface(name=Bounds for - com.yitec.tryon/com.unity3d.player.UnityPlayerActivity@0)/@0x1e369b9 frame = 8
[DP] dp(1) 1 android.view.SurfaceView.updateSurface:1311 android.view.SurfaceView.setFrame:646 android.view.View.layout:24472 
[DP] pdf(0) 1 android.view.SurfaceView.notifyDrawFinished:577 android.view.SurfaceView.performDrawFinished:564 android.view.SurfaceView.lambda$TWz4D2u33ZlAmRtgKzbqqDue3iM:0 
[DP] rdf()
ViewPostIme pointer 0
ViewPostIme pointer 1
WebCamDevice Changed: Camera 0
WebCamDeviceSelectorController:OnValueChanged(Dropdown)
<Start>d__2:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

WebCamTexture Graphics Format: 88
<ResetScreen>d__16:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
SceneDirector:ResetCamera(Nullable`1)
SceneDirector:ChangeWebCamDevice(Nullable`1)
WebCamDeviceSelectorController:OnValueChanged(Dropdown)
<Start>d__2:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

EGL Context Found (503682402736)
SceneDirector:SetupGpuResources()
SceneDirector:StartGraph()
SceneDirector:ChangeWebCamDevice(Nullable`1)
WebCamDeviceSelectorController:OnValueChanged(Dropdown)
<Start>d__2:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

hand_landmark is saved to /storage/emulated/0/Android/data/com.yitec.tryon/files/Cache/hand_landmark.bytes (length=3792620)
Mediapipe.AssetBundleManager:WriteCacheFile(TextAsset, String, Boolean)
HandTrackingGraph:PrepareDependentAssets()
DemoGraph:Initialize()
DemoGraph:Initialize(GpuResources, GlCalculatorHelper)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

hand_recrop is saved to /storage/emulated/0/Android/data/com.yitec.tryon/files/Cache/hand_recrop.bytes (length=123792)
Mediapipe.AssetBundleManager:WriteCacheFile(TextAsset, String, Boolean)
HandTrackingGraph:PrepareDependentAssets()
DemoGraph:Initialize()
DemoGraph:Initialize(GpuResources, GlCalculatorHelper)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

handedness is saved to /storage/emulated/0/Android/data/com.yitec.tryon/files/Cache/handedness.txt (length=11)
Mediapipe.AssetBundleManager:WriteCacheFile(TextAsset, String, Boolean)
HandTrackingGraph:PrepareDependentAssets()
DemoGraph:Initialize()
DemoGraph:Initialize(GpuResources, GlCalculatorHelper)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

palm_detection is saved to /storage/emulated/0/Android/data/com.yitec.tryon/files/Cache/palm_detection.bytes (length=3877888)
Mediapipe.AssetBundleManager:WriteCacheFile(TextAsset, String, Boolean)
DemoGraph:Initialize()
DemoGraph:Initialize(GpuResources, GlCalculatorHelper)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

Loaded dependent assets
DemoGraph:Initialize()
DemoGraph:Initialize(GpuResources, GlCalculatorHelper)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

ViewPostIme pointer 0
ViewPostIme pointer 1
ViewPostIme key 0
ViewPostIme key 1
ViewPostIme pointer 0
ViewPostIme pointer 1
HANDEDNESS:Right
HandTrackingAnnotationController:Draw(Transform, List`1, List`1, List`1, List`1, Boolean)
HandTrackingGraph:RenderOutput(WebCamScreenController, TextureFrame)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

HANDEDNESS:Right
HandTrackingAnnotationController:Draw(Transform, List`1, List`1, List`1, List`1, Boolean)
HandTrackingGraph:RenderOutput(WebCamScreenController, TextureFrame)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

ViewPostIme key 0
ViewPostIme key 1
HANDEDNESS:Left
HandTrackingAnnotationController:Draw(Transform, List`1, List`1, List`1, List`1, Boolean)
HandTrackingGraph:RenderOutput(WebCamScreenController, TextureFrame)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

ViewPostIme pointer 0
ViewPostIme pointer 1
ViewPostIme pointer 0
ViewPostIme pointer 1
WebCamDevice Changed: Camera 0
WebCamDeviceSelectorController:OnValueChanged(Dropdown)
<Start>d__2:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

WebCamTexture Graphics Format: 88
<ResetScreen>d__16:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
SceneDirector:ResetCamera(Nullable`1)
SceneDirector:ChangeWebCamDevice(Nullable`1)
WebCamDeviceSelectorController:OnValueChanged(Dropdown)
<Start>d__2:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

EGL Context Found (503682402736)
SceneDirector:SetupGpuResources()
SceneDirector:StartGraph()
SceneDirector:ChangeWebCamDevice(Nullable`1)
WebCamDeviceSelectorController:OnValueChanged(Dropdown)
<Start>d__2:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

The AssetBundle 'mediapipe' can't be loaded because another AssetBundle with the same files is already loaded.
Mediapipe.AssetBundleManager:LoadAssetBundle()
Mediapipe.AssetBundleManager:PrepareAsset(String, String, Boolean)
HandTrackingGraph:PrepareDependentAssets()
DemoGraph:Initialize()
DemoGraph:Initialize(GpuResources, GlCalculatorHelper)
<RunGraph>d__25:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)

[ line -1123386392]

NullReferenceException: Object reference not set to an instance of an object.
  at Mediapipe.AssetBundleManager.PrepareAsset (System.String name, System.String uniqueKey, System.Boolean overwrite) [0x00000] in <00000000000000000000000000000000>:0 
  at HandTrackingGraph.PrepareDependentAssets () [0x00000] in <00000000000000000000000000000000>:0 
  at DemoGraph.Initialize () [0x00000] in <00000000000000000000000000000000>:0 
  at DemoGraph.Initialize (Mediapipe.GpuResources gpuResources, Mediapipe.GlCalculatorHelper gpuHelper) [0x00000] in <00000000000000000000000000000000>:0 
  at SceneDirector+<RunGraph>d__25.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <00000000000000000000000000000000>:0

**Comment by ROBYER1** [comment]
I have the fixes for this, in Webcamscreencontroller, add
`    private void OnApplicationQuit()
    {
        webCamTexture.Stop();
        webCamTexture = null;
        pixelData = null;
    }
    private void OnDisable()
    {
        if (webCamTexture != null)
        {
            webCamTexture.Stop();
            webCamTexture = null;
        }

        pixelData = null;
    }`

Then also swap these files out in your project to see if it works, if not then look at what I did in them, I just made sure to unloadassetbundles when the scene is left so they can be reloaded
[tofix.zip](https://github.com/homuler/MediaPipeUnityPlugin/files/7020824/tofix.zip)

**Comment by quangminhdinh** [comment]
Thank you, it works perfectly.

**Comment by ROBYER1** [comment]
@homuler rather than me making a new issue for this, I was noticing crashes in editor/build when reloading scenes until I made these changes:

1. Disable Glog in Bootstrap.cs
2. In ResourceManager.cs, comment out line 36 as `                    //throw new InvalidOperationException("ResourceManager can be initialized only once");
`
3. In WebCamSource.cs, add the following lines to clean up WebcamTexture on destroy:
```
        private void OnDestroy()
        {
            if (_webCamTexture != null)
            {
                _webCamTexture.Stop();
                _webCamTexture = null;
            }
        }
        private void OnApplicationQuit()
        {
            _webCamTexture.Stop();
            _webCamTexture = null;
        }
```

Is this the correct way to deal with reloading a scene? We are only using Hand tracking across scenes and by doing this it works fine.

**Comment by homuler** [comment]
@ROBYER1 
> Is this the correct way to deal with reloading a scene?

The sample scenes (and code under `Assets/MediaPipeUnity`) are implemented for demonstration purposes and not intended to be reused in different contexts, so if you're reloading the scene in a different way than the sample app does, it would be wrong.

With that in mind, in order to reuse the code, you should note that what `Bootstrap` does must be executed only once (cf. https://github.com/homuler/MediaPipeUnityPlugin/issues/446#issuecomment-1033793353).
As for `WebCamSource`, I think you should stop it before reloading the scene.
